ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java

```diff
@@ -63,4 +111,69 @@ private boolean isValid(AppendEntriesReplyProto reply) {
       return appender.getCallIdComparator().compare(reply.getServerReply().getCallId(), minCallId) >= 0;
     }
   }
+
+  static class HeartbeatBroadcastListener {
+    private final Map<RaftPeerId, HeartbeatAck> pendingAcknowledgements;
+    private final int groupPeerCount;
+    private int ack;
+    private volatile boolean done;
+    private final CompletableFuture<Boolean> result;
+
+    HeartbeatBroadcastListener(List<LogAppender> senders) {
+      this.groupPeerCount = senders.size() + 1;
+      this.ack = 0;
+      this.done = false;
+      result = new CompletableFuture<>();
+
+      pendingAcknowledgements = new ConcurrentHashMap<>();
+      senders.forEach(sender ->
+          pendingAcknowledgements.put(sender.getFollowerId(), new HeartbeatAck(sender)));
+    }
+
+    private void handleReply(RaftPeerId peerId, AppendEntriesReplyProto reply) {
+      if (done) {
+        return;
+      }
+
+      Optional.ofNullable(pendingAcknowledgements.get(peerId)).ifPresent(heartbeatAck -> {
+        boolean firstAcknowledged = heartbeatAck.receive(reply);
+        if (firstAcknowledged) {
+          onHeartbeatAcknowledged();
+        }
+      });
+    }
+
+    private synchronized void handleTimeout() {
+      if (!done) {
+        done = true;
+        result.complete(false);
+      }
+    }
+
+    private boolean isDone() {
+      return this.done;
+    }
+
+    synchronized void onHeartbeatAcknowledged() {
+      if (done) {
+        return;
+      }
+
+      ack++;
+
+      if (isMajorityAck()) {
+        done = true;
+        result.complete(true);
+      }
+    }
+
+    private synchronized boolean isMajorityAck() {
+      // include leader itself
+      return ack + 1 > groupPeerCount / 2;
+    }
```

Comment:
The configuration may change so that we should use the same logic in `LeaderStateImpl.getMajorityMin` to calculate majority.  We may add the following method:
```java
//LeaderStateImpl.java
  private boolean hasMajority(Predicate<RaftPeerId> isAcked) {
    final RaftPeerId selfId = server.getId();
    final RaftConfigurationImpl conf = server.getRaftConf();

    final List<RaftPeerId> followers = voterLists.get(0);
    final boolean includeSelf = conf.containsInConf(selfId);
    final boolean newConf = hasMajority(isAcked, followers, includeSelf);

    if (!conf.isTransitional()) {
      return newConf;
    } else { // configuration is in transitional state
      final List<RaftPeerId> oldFollowers = voterLists.get(1);
      final boolean includeSelfInOldConf = conf.containsInOldConf(selfId);
      final boolean oldConf = hasMajority(isAcked, oldFollowers, includeSelfInOldConf);
      return newConf && oldConf;
    }
  }
```

Reply:
done